### PR TITLE
SC-16350: More specific errors

### PR DESCRIPTION
**Pull request by fsdvh**
We're trying to find a way to show users on UI some meaningful errors. The problem is that we return plan string error messages as a response from a ballista. 

I propose to reserve `DataFusionError::External` for errors that can be shown in UI, then we can propagate such knowledge back to our API server and then back to the user.

In this PR I'm attempting to solve the issue of untyped, unclassified errors in ballista without introducing lots of chaos in the codebase

Using @ch-sc rebase branch as a base for now, to reduce the amount of noise

Any suggestions are welcome!

### CODE REVIEW

**Review by Dandandan — APPROVED** [review]
Overall looks good to me

File: ballista/scheduler/src/cluster/mod.rs [hunk 1]
```diff
@@ -320,6 +322,42 @@ pub enum JobStateEvent {
 /// by any schedulers with a shared `ClusterState`
 pub type JobStateEventStream = Pin<Box<dyn Stream<Item = JobStateEvent> + Send>>;
 
+#[derive(Clone)]
+pub enum FailureReason {
```
[the remark this answers is not in the document]
Comment (fsdvh) [reply ballista/scheduler/src/cluster/mod.rs:326]: Yes, I agree. I tried to propose the path of least changes for the desired feature, while the comprehensive solution would be to return the error itself, with your blessing I will serialize BallistaError to protobuf 😄

File: ballista/core/proto/datafusion.proto [hunk 2]
```diff
@@ -21,8 +21,8 @@ syntax = "proto3";
 package datafusion;
 
 option java_multiple_files = true;
-option java_package = "org.apache.arrow.datafusion.protobuf";
```
Comment (fsdvh) [inline ballista/core/proto/datafusion.proto:24]: all changes in this file are the result of: `buf fmt -w`

File: ballista/scheduler/proto/keda.proto [hunk 3]
```diff
@@ -19,45 +19,46 @@
 syntax = "proto3";
```
Comment (fsdvh) [inline ballista/scheduler/proto/keda.proto:20]: same here

**Review by thinkharderdev — APPROVED** [review]
Nice!

**Pull request merged**